File: Estudos solo/Python/LIFO.py
STACKMAX = 100

class LIFO:
    def __init__ (self):
        self.top = -1
        self.data = []

    def empty(self):
        if not self.data:
            print("Pilha vazia")
            return True
        return False
    
    def push(self, valor):
        if len(self.data) < STACKMAX:
            self.data.append(valor)
            self.top = len(self.data) - 1
        else:
            return print("Pilha cheia")
        
    def pop(self):
        if self.empty():
            return None
        
        self.data.pop()
        self.top = len(self.data) - 1
        return 1
    
    def stacktop(self):
        if self.empty():
            return None
        print(f"Topo: {self.data[-1]}")

    def stackprint(self):
        if not self.empty():
            for item in reversed(self.data):
                print(f"{item}  ")

pilha = LIFO()
pilha.top = -1

pilha.empty()

pilha.push(2)
pilha.push(99)
pilha.push(67)
pilha.push(66)

pilha.stacktop()
pilha.stackprint()

pilha.pop()

pilha.stacktop()
pilha.stackprint()

#Last in, First Out: primeiro a sair foi o ultimo que entrou (pilha)

File: Estudos solo/LIFO.cpp
#include <stdio.h>
#include <stdlib.h>

#define MAXSTACK 100

struct stack {
    int top;
    int data[MAXSTACK];
};

bool empty(struct stack *pnt){
    if(pnt->top == -1){
        printf("\nPilha vazia");
        return (1);
    }
    return (0);
}

void stacktop(struct stack *pnt){
    if(empty(pnt)) return;
    printf("\n%d e o topo da pilha", pnt->data[pnt->top]);
    return;
}

void stackPrint(struct stack *pnt){
    if(empty(pnt)) return;
    printf("\n");
    for (int i = 0; i<= pnt->top; i++){
        printf("%d   ", pnt->data[i]);
    }
    return;
}

void push(struct stack *pnt, int valor){
    if(pnt->top < MAXSTACK){
        pnt->top++;
        pnt->data[pnt->top] = valor;
    }
    return;
}

void pop(struct stack *pnt){
    if(empty(pnt)) return;
    pnt->top--;
    return;
}

int main(){
    struct stack s;
    s.top = -1;

    
    stacktop(&s);

    push(&s, 2);
    push(&s, 67);
    push(&s, 16);
    
    stackPrint(&s);

    stacktop(&s);

    pop(&s);

    stackPrint(&s);
    
    stacktop(&s);

    return 0;
}